src/button/button-xbm.c

```diff
@@ -278,7 +271,9 @@ button_xbm_load(const char *filename, struct lab_data_buffer **buffer,
 	color = u32(rgba);
 
 	/* Read file into memory as it's easier to tokenzie that way */
-	char *token_buffer = grab_file(xbm_path(filename));
+	char filename[4096] = { 0 };
+	button_filename(button_name, filename, sizeof(filename));
```

Comment:
Just a suggestion: can we start using `ARRAY_SIZE()` instead of `sizeof()` when array len is involved?

Reply:
Yes, I'm up for that.
IIRC we started off like that and then read the Devault coding style and decided to follow.

I think we even disabled it in checkpatch.pl so if we turn it back on we should find the places to tweak.

Can't grep right now, but think we even define it locally somewhere.

Question is, do we go with a simple version or one that does type checking (which I think Linux does). Not sure if the latter need GNU_SOURCE which we've so far avoided.

Reply:
That's the kernel one:
```
#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof((arr)[0]) + __must_be_array(arr))
```

Magic happens here:
```
#define __must_be_array(a)	BUILD_BUG_ON_ZERO(__same_type((a), &(a)[0]))
```

And it relies on `__builtin_types_compatible_p()`:
```

#define __same_type(a, b) __builtin_types_compatible_p(typeof(a), typeof(b))
```

Both GCC and clang support this builtin. If you have any suspicions about some system that does not have this builtin, we can introduce #ifdef guard.

Reply:
See #1086

Reply:
Using `ARRAY_SIZE` for char arrays feels wrong to me.

Reply:
Agree. @heroin-moose - is ARRAY_SIZE used for char[123] in the Linux Kernel?

Reply:
For any array.

Reply:
```
	u8 data[4];
[...]
	ret = i2c_smbus_write_i2c_block_data(chip->client,
					BH1770_ALS_TH_UP_0,
					ARRAY_SIZE(data),
					data);
```

Reply:
So in the kernel would you not use `sizeof()` in this example:

```
char filename[4096] = { 0 };
button_filename(button_name, filename, sizeof(filename));
```